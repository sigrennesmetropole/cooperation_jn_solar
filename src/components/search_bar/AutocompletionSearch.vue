<script setup lang="ts">
import { computed, ref, watch } from 'vue'
import iconSearchEmpty from '@/assets/icons/search-empty.svg'
import type {
  AutoCompletion,
  AddressRva,
  AddressOrganization,
  AddressCommune,
  AddressStreet,
} from '@/model/address.model'

const props = defineProps<{
  isDisplayAutocompletion: boolean
  isDisplayFilters: boolean
  isEmptySearch: boolean
  highlightedAutocompletion: AutoCompletion
}>()

const emit = defineEmits(['goToAddress'])
const container = ref<HTMLDivElement | null>(null)

const autocompletionFormatted = computed(() => {
  return [
    {
      type: 'rva',
      name_key_id: 'idaddress',
      name_key_address: 'addr3',
      name_singular: 'Adresse',
      name_plural: 'Adresses',
      data: props.highlightedAutocompletion.addressRva,
    },
    {
      type: 'organization',
      name_key_id: 'id',
      name_key_address: 'addr',
      name_singular: 'Organisme',
      name_plural: 'Organismes',
      data: props.highlightedAutocompletion.addressOrganization,
    },
    {
      type: 'streets',
      name_key_id: 'idlane',
      name_key_address: 'name4',
      name_singular: 'Voie',
      name_plural: 'Voies',
      data: props.highlightedAutocompletion.streets,
    },
    {
      type: 'communes',
      name_key_id: 'nameindex',
      name_key_address: 'name',
      name_singular: 'Commune',
      name_plural: 'Communes',
      data: props.highlightedAutocompletion.communes,
    },
  ]
})

const getAddressValue = (
  item2: AddressRva | AddressOrganization | AddressCommune | AddressStreet,
  key: string
) => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  return (item2 as any)[key]
}

const getIdValue = (
  item2: AddressRva | AddressOrganization | AddressCommune | AddressStreet,
  key: string
) => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  return (item2 as any)[key]
}

const selectedIndex = ref(-1)
const scrollToSelectedItem = () => {
  if (container.value && getSelectedIndex() !== -1) {
    const { categoryIndex, localIndex } = getCategoryAndLocalIndex(
      getSelectedIndex()
    )
    const selectedItem = container.value.querySelector(
      `.autocompletion-item[data-category-index="${categoryIndex}"][data-item-index="${localIndex}"]`
    ) as HTMLElement

    if (selectedItem) {
      container.value.scrollTop =
        selectedItem.offsetTop - container.value.clientHeight / 2
    }
  }
}

const onKeyPress = (event: KeyboardEvent) => {
  if (event.key === 'ArrowUp') {
    event.preventDefault()
    if (getSelectedIndex() > 0) {
      selectedIndex.value--
    }
    scrollToSelectedItem()
  } else if (event.key === 'ArrowDown') {
    event.preventDefault()
    const totalItems = autocompletionFormatted.value.reduce(
      (sum, item) => sum + item.data.length,
      0
    )
    if (getSelectedIndex() < totalItems - 1) {
      selectedIndex.value++
    } else if (getSelectedIndex() >= totalItems - 1) {
      selectedIndex.value = 0
    }
    scrollToSelectedItem()
  } else if (event.key === 'Enter') {
    const { categoryIndex, localIndex } = getCategoryAndLocalIndex(
      getSelectedIndex()
    )
    const item = autocompletionFormatted.value[categoryIndex]
    const selectedItem = item.data[localIndex]
    emit('goToAddress', {
      item: selectedItem,
      type: item.type,
      addr: getAddressValue(selectedItem, item.name_key_address),
    })
  }
}

watch(
  () => props.isDisplayAutocompletion,
  (value) => {
    if (value) {
      window.addEventListener('keydown', onKeyPress)
    } else {
      window.removeEventListener('keydown', onKeyPress)
    }
  },
  { immediate: true }
)

const getGlobalIndex = (categoryIndex: number, itemIndex: number) => {
  let globalIndex = 0
  for (let i = 0; i < categoryIndex; i++) {
    globalIndex += autocompletionFormatted.value[i].data.length
  }
  return globalIndex + itemIndex
}

const getCategoryAndLocalIndex = (globalIndex: number) => {
  let categoryIndex = 0
  let localIndex = globalIndex

  while (
    categoryIndex < autocompletionFormatted.value.length &&
    localIndex >= autocompletionFormatted.value[categoryIndex].data.length
  ) {
    localIndex -= autocompletionFormatted.value[categoryIndex].data.length
    categoryIndex++
  }

  return { categoryIndex, localIndex }
}

const getSelectedIndex = () => {
  return selectedIndex.value
}
</script>

<template>
  <div
    ref="container"
    class="font-dm-sans flex flex-col rounded mt-0 shadow-lg bg-white max-h-[300px] overflow-auto w-[402px] scrollbar scrollbar-thumb-black scrollbar-track-inerth scrollbar-thumb-rounded-full scrollbar-w-1"
    v-if="isDisplayAutocompletion && !isDisplayFilters"
  >
    <template v-for="(item, index) in autocompletionFormatted" :key="index">
      <ul v-if="item.data.length > 0">
        <li class="border-b border-neutral-200 py-1">
          <span class="font-dm-sans font-normal text-xs text-neutral-600 px-3">
            <template v-if="item.data.length > 1">
              {{ item.name_plural }}
            </template>
            <template v-else> {{ item.name_singular }} </template>
          </span>
        </li>
        <li
          v-for="(item2, index2) in item.data"
          :key="getIdValue(item2, item.name_key_id)"
          @click="
            $emit('goToAddress', {
              item: item2,
              type: item.type,
              addr: getAddressValue(item2, item.name_key_address),
            })
          "
          class="cursor-pointer border-b border-neutral-200 py-1 autocompletion-item"
          @mouseover="selectedIndex = getGlobalIndex(index, index2)"
          :class="{
            'bg-neutral-100': selectedIndex === getGlobalIndex(index, index2),
          }"
          :data-category-index="index"
          :data-item-index="index2"
        >
          <div
            v-html="getAddressValue(item2, item.name_key_address)"
            class="px-3 font-dm-sans font-normal text-sm"
          ></div>
        </li>
      </ul>
    </template>
  </div>
  <div
    v-else-if="props.isEmptySearch"
    class="flex flex-row items-center rounded px-3 py-4 mt-0 shadow-lg bg-white"
  >
    <img :src="iconSearchEmpty" class="w-4 h-4" />
    <span class="text-base font-medium font-dm-sans ml-2">
      Aucun r√©sultat
    </span>
  </div>
</template>
