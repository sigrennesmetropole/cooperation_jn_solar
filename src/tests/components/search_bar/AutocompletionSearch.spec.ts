// @ts-nocheck
import AutocompletionSearch from '@/components/search_bar/AutocompletionSearch.vue'
import { describe, it, expect, beforeEach } from 'vitest'
import { mount, VueWrapper } from '@vue/test-utils'
import highlightedAutocompletion from './highlightedAutocompletion.json'

describe('AutocompletionSearch.vue', () => {
  let wrapper: VueWrapper

  beforeEach(async () => {
    wrapper = mount(AutocompletionSearch, {
      propsData: {
        isDisplayAutocompletion: true,
        isDisplayFilters: false,
        isEmptySearch: false,
        highlightedAutocompletion,
      },
    })
  })

  it('renders correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders the correct number of autocompletion items', () => {
    const autocompletionItems = wrapper.findAll('.autocompletion-item')
    expect(autocompletionItems.length).toBe(14)
  })

  it('emits the goToAddress event when an item is clicked', async () => {
    const autocompletionItems = wrapper.findAll('.autocompletion-item')
    await autocompletionItems[0].trigger('click')

    const goToAddressEvent = wrapper.emitted('goToAddress')
    expect(goToAddressEvent).toBeTruthy()
    expect(goToAddressEvent![0]).toEqual([
      {
        item: {
          insee: '35131',
          idlane: '3179',
          idaddress: '87781',
          number: '37',
          extension: '',
          building: 'B',
          addr1: '37 B',
          addr2: '37 B rue de Rennes',
          addr3: "37 B rue de <strong>Rennes</strong>, l'Hermitage",
          x: '-1.80428706975102',
          y: '48.1233099744015',
        },
        type: 'rva',
        addr: "37 B rue de <strong>Rennes</strong>, l'Hermitage",
      },
    ])
  })

  it('updates selectedIndex when mouse is over an item', async () => {
    const autocompletionItems = wrapper.findAll('.autocompletion-item')

    await autocompletionItems[0].trigger('mouseover')
    const selectedIndex = wrapper.vm.getSelectedIndex()
    expect(selectedIndex).toBe(0)
    expect(autocompletionItems[0].classes()).toContain('bg-neutral-100')

    await autocompletionItems[10].trigger('mouseover')
    const selectedIndex2 = wrapper.vm.getSelectedIndex()
    expect(selectedIndex2).toBe(10)
    expect(autocompletionItems[10].classes()).toContain('bg-neutral-100')
  })

  it('handles key up event correctly', async () => {
    //Set the selected index to 10
    const autocompletionItems = wrapper.findAll('.autocompletion-item')
    await autocompletionItems[10].trigger('mouseover')

    //Test the key up event
    const initialSelectedIndex = wrapper.vm.getSelectedIndex()
    wrapper.vm.onKeyPress(new KeyboardEvent('keydown', { key: 'ArrowUp' }))
    const selectedIndex = wrapper.vm.getSelectedIndex()
    expect(selectedIndex).toBe(Math.max(initialSelectedIndex - 1, 0))
  })

  it('handles key down event correctly', async () => {
    //Set the selected index to 10
    const autocompletionItems = wrapper.findAll('.autocompletion-item')
    await autocompletionItems[10].trigger('mouseover')

    //Test the key down event
    const initialSelectedIndex = wrapper.vm.getSelectedIndex()
    wrapper.vm.onKeyPress(new KeyboardEvent('keydown', { key: 'ArrowDown' }))
    const selectedIndex = wrapper.vm.getSelectedIndex()
    expect(selectedIndex).toBe(initialSelectedIndex + 1)
  })

  it('handles enter event correctly', async () => {
    //Set the selected index to 10
    const autocompletionItems = wrapper.findAll('.autocompletion-item')
    await autocompletionItems[10].trigger('mouseover')

    // Trigger the 'Enter' keydown event
    wrapper.vm.onKeyPress(new KeyboardEvent('keydown', { key: 'Enter' }))

    // Check if the 'goToAddress' event has been emitted
    const emittedEvents = wrapper.emitted()
    expect(emittedEvents.goToAddress).toBeTruthy()

    // Check the specific parameters passed to the 'goToAddress' event
    const eventData = emittedEvents.goToAddress[0][0]
    expect(eventData).toMatchObject({
      item: {
        insee: '35250',
        idlane: '1079',
        fantoir: '0022',
        type: 'Rue',
        name: 'Rue de Rennes',
        name2: 'RUE DE RENNES',
        name3: 'de Rennes',
        nameindex: 'Rennes (Rue de)',
        name4: 'Rue de <strong>Rennes</strong>, Saint-Armel',
        lowerCorner: '-1.6002134670577 48.0140820822421',
        upperCorner: '-1.59504190227835 48.0149596383537',
      },
      type: 'streets',
      addr: 'Rue de <strong>Rennes</strong>, Saint-Armel',
    })
  })
})
