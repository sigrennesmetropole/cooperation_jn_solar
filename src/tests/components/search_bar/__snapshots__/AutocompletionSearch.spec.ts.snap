// Vitest Snapshot v1

exports[`AutocompletionSearch.vue > renders correctly 1`] = `
"<div class=\\"font-dm-sans flex flex-col rounded mt-0 shadow-lg bg-white max-h-[300px] overflow-auto w-[402px] scrollbar scrollbar-thumb-black scrollbar-track-inerth scrollbar-thumb-rounded-full scrollbar-w-1\\">
  <ul>
    <li class=\\"border-b border-neutral-200 py-1\\"><span class=\\"font-dm-sans font-normal text-xs text-neutral-600 px-3\\">Adresses</span></li>
    <li class=\\"cursor-pointer border-b border-neutral-200 py-1 autocompletion-item\\" data-category-index=\\"0\\" data-item-index=\\"0\\">
      <div class=\\"px-3 font-dm-sans font-normal text-sm\\">37 B rue de <strong>Rennes</strong>, l'Hermitage</div>
    </li>
    <li class=\\"cursor-pointer border-b border-neutral-200 py-1 autocompletion-item\\" data-category-index=\\"0\\" data-item-index=\\"1\\">
      <div class=\\"px-3 font-dm-sans font-normal text-sm\\">36 rue de <strong>Rennes</strong>, Chevaigné</div>
    </li>
    <li class=\\"cursor-pointer border-b border-neutral-200 py-1 autocompletion-item\\" data-category-index=\\"0\\" data-item-index=\\"2\\">
      <div class=\\"px-3 font-dm-sans font-normal text-sm\\">32 rue de <strong>Rennes</strong>, Corps-Nuds</div>
    </li>
    <li class=\\"cursor-pointer border-b border-neutral-200 py-1 autocompletion-item\\" data-category-index=\\"0\\" data-item-index=\\"3\\">
      <div class=\\"px-3 font-dm-sans font-normal text-sm\\">37 A rue de <strong>Rennes</strong>, l'Hermitage</div>
    </li>
    <li class=\\"cursor-pointer border-b border-neutral-200 py-1 autocompletion-item\\" data-category-index=\\"0\\" data-item-index=\\"4\\">
      <div class=\\"px-3 font-dm-sans font-normal text-sm\\">32 rue de <strong>Rennes</strong>, Chevaigné</div>
    </li>
  </ul>
  <ul>
    <li class=\\"border-b border-neutral-200 py-1\\"><span class=\\"font-dm-sans font-normal text-xs text-neutral-600 px-3\\">Organismes</span></li>
    <li class=\\"cursor-pointer border-b border-neutral-200 py-1 autocompletion-item\\" data-category-index=\\"1\\" data-item-index=\\"0\\">
      <div class=\\"px-3 font-dm-sans font-normal text-sm\\">Aéroport <strong>Rennes</strong> Bretagne</div>
    </li>
    <li class=\\"cursor-pointer border-b border-neutral-200 py-1 autocompletion-item\\" data-category-index=\\"1\\" data-item-index=\\"1\\">
      <div class=\\"px-3 font-dm-sans font-normal text-sm\\">Agence Locale de l'Energie et du Climat du Pays de <strong>Rennes</strong></div>
    </li>
    <li class=\\"cursor-pointer border-b border-neutral-200 py-1 autocompletion-item\\" data-category-index=\\"1\\" data-item-index=\\"2\\">
      <div class=\\"px-3 font-dm-sans font-normal text-sm\\">Aide à Domicile de <strong>Rennes</strong></div>
    </li>
  </ul>
  <ul>
    <li class=\\"border-b border-neutral-200 py-1\\"><span class=\\"font-dm-sans font-normal text-xs text-neutral-600 px-3\\">Voies</span></li>
    <li class=\\"cursor-pointer border-b border-neutral-200 py-1 autocompletion-item\\" data-category-index=\\"2\\" data-item-index=\\"0\\">
      <div class=\\"px-3 font-dm-sans font-normal text-sm\\">Chemin d'exploitation 46 des Ga<strong>rennes</strong>, Bruz</div>
    </li>
    <li class=\\"cursor-pointer border-b border-neutral-200 py-1 autocompletion-item\\" data-category-index=\\"2\\" data-item-index=\\"1\\">
      <div class=\\"px-3 font-dm-sans font-normal text-sm\\">Rue de <strong>Rennes</strong>, Thorigné-Fouillard</div>
    </li>
    <li class=\\"cursor-pointer border-b border-neutral-200 py-1 autocompletion-item\\" data-category-index=\\"2\\" data-item-index=\\"2\\">
      <div class=\\"px-3 font-dm-sans font-normal text-sm\\">Rue de <strong>Rennes</strong>, Saint-Armel</div>
    </li>
    <li class=\\"cursor-pointer border-b border-neutral-200 py-1 autocompletion-item\\" data-category-index=\\"2\\" data-item-index=\\"3\\">
      <div class=\\"px-3 font-dm-sans font-normal text-sm\\">Rue de <strong>Rennes</strong>, Acigné</div>
    </li>
    <li class=\\"cursor-pointer border-b border-neutral-200 py-1 autocompletion-item\\" data-category-index=\\"2\\" data-item-index=\\"4\\">
      <div class=\\"px-3 font-dm-sans font-normal text-sm\\">Rue de <strong>Rennes</strong>, l'Hermitage</div>
    </li>
  </ul>
  <ul>
    <li class=\\"border-b border-neutral-200 py-1\\"><span class=\\"font-dm-sans font-normal text-xs text-neutral-600 px-3\\">Commune</span></li>
    <li class=\\"cursor-pointer border-b border-neutral-200 py-1 autocompletion-item\\" data-category-index=\\"3\\" data-item-index=\\"0\\">
      <div class=\\"px-3 font-dm-sans font-normal text-sm\\"><strong>Rennes</strong></div>
    </li>
  </ul>
</div>"
`;
